subprojects/language-groovy/src/main/java/org/gradle/api/tasks/compile/GroovyCompile.java

```diff
@@ -137,11 +206,177 @@ private void warnIfCompileAvoidanceEnabled() {
+                    }
+                });
+            } else {
+                compiler = cleaningGroovyCompiler;
+            }
         }
         return compiler;
     }
 
+    /**
+     * The sources for incremental change detection.
+     *
+     * @since 5.6
+     */
+    @Incubating
+    @SkipWhenEmpty
+    @PathSensitive(PathSensitivity.RELATIVE) // Java source files are supported, too. Therefore we should care about the relative path.
+    @InputFiles
+    protected FileCollection getStableSources() {
+        return stableSources;
+    }
+
+    /**
+     * Injects and returns an instance of {@link FileSystemSnapshotter}.
+     *
+     * @since 5.6
+     */
+    @Inject
+    protected FileSystemSnapshotter getFileSystemSnapshotter() {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Injects and returns an instance of {@link BuildOperationExecutor}.
+     *
+     * @since 5.6
+     */
+    @Inject
+    protected BuildOperationExecutor getBuildOperationExecutor() {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Injects and returns an instance of {@link FileOperations}.
+     * @since 5.6
+     */
+    @Inject
+    protected FileOperations getFileOperations() {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Injects and returns an instance of {@link StreamHasher}.
+     * @since 5.6
+     */
+    @Inject
+    protected StreamHasher getStreamHasher() {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Injects and returns an instance of {@link FileHasher}.
+     * @since 5.6
+     */
+    @Inject
+    protected FileHasher getFileHasher() {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Injects and returns an instance of {@link StringInterner}.
+     * @since 5.6
+     */
+    @Inject
+    protected StringInterner getStringInterner() {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Injects and returns an instance of {@link GeneralCompileCaches}.
+     * @since 5.6
+     */
+    @Inject
+    protected GeneralCompileCaches getCompileCaches() {
+        throw new UnsupportedOperationException();
+    }
+
+    private static class IncrementalInputsAdapter implements IncrementalTaskInputs {
+        private final InputChanges inputChanges;
+        private final FileCollection sources;
+
+        public IncrementalInputsAdapter(InputChanges inputChanges, FileCollection sources) {
+            this.inputChanges = inputChanges;
+            this.sources = sources;
+        }
+
+        @Override
+        public boolean isIncremental() {
+            return inputChanges != null && inputChanges.isIncremental();
+        }
+
+        @Override
+        public void outOfDate(Action<? super InputFileDetails> outOfDateAction) {
+            for (final FileChange change : inputChanges.getFileChanges(sources)) {
```

Comment:
You also need to iterate the file changes for the classpath (and for every other incremental input property). If you don't, the changes will be ignored. I wonder why no test failed...